### PR TITLE
Rewrite more functions

### PR DESCRIPTION
**Pull request by systay**
This PR adds support for `SCHEMA()` and `FOUND_ROWS()`. `SQL_CALC_FOUND_ROWS` is being implemented in a separate PR

### CODE REVIEW

**Comment by morgo** [comment]
There is a bug in `SQL_CALC_FOUND_ROWS`. Here is the test case:
```
DROP TABLE IF EXISTS t1;
CREATE TABLE t1 (id INT NOT NULL PRIMARY KEY auto_increment);
INSERT INTO t1 VALUES (NULL),(NULL),(NULL),(NULL),(NULL);

SELECT SQL_CALC_FOUND_ROWS * FROM t1 LIMIT 1;
SELECT FOUND_ROWS();

SELECT SQL_CALC_FOUND_ROWS * FROM t1 LIMIT 2;
SELECT FOUND_ROWS();
```

The first `FOUND_ROWS()` returns 1, second returns 2. Both would be correct if the `SQL_CALC_FOUND_ROWS` was not present. Because it is, both calls are expected to return 5.

**Comment by systay** [comment]
> There is a bug in `SQL_CALC_FOUND_ROWS`.

Well, I've only added parsing support. I have yet to add the bugs to the planner. ;)

File: go/vt/vtgate/executor.go [hunk 1]
```diff
@@ -1444,7 +1451,7 @@ func (e *Executor) getPlan(vcursor *vcursorImpl, sql string, comments sqlparser.
 	if result, ok := e.plans.Get(planKey); ok {
```
Comment (sougou) [inline go/vt/vtgate/executor.go]: Let's rename this to `plan`. Otherwise it's confusing w.r.t. the other result.

File: go/vt/vtgate/executor.go [hunk 2]
```diff
@@ -145,6 +145,9 @@ func (e *Executor) Execute(ctx context.Context, method string, safeSession *Safe
 
 	logStats := NewLogStats(ctx, method, sql, bindVars)
 	result, err = e.execute(ctx, safeSession, sql, bindVars, logStats)
+	if err == nil {
+		safeSession.FoundRows = result.RowsAffected
+	}
```
Comment (harshit-gangal) [inline go/vt/vtgate/executor.go]: do you not need locking around updating the foundRows?

File: go/vt/vtgate/engine/primitive.go [hunk 3]
```diff
@@ -94,10 +96,8 @@ type Plan struct {
 	Rows uint64 `json:",omitempty"`
 	// Total number of errors
 	Errors uint64 `json:",omitempty"`
-	// NeedsLastInsertID signals whether this plan will need to be provided with last_insert_id
-	NeedsLastInsertID bool `json:"-"` // don't include in the json representation
-	// NeedsDatabaseName signals whether this plan will need to be provided with the database name
-	NeedsDatabaseName bool `json:"-"` // don't include in the json representation
+
+	sqlparser.BindVarNeeds
```
Comment (harshit-gangal) [inline go/vt/vtgate/engine/primitive.go]: nit: add comment for being in the plan like others. Also need to add json:"-" if this does not need json representation.

File: go/vt/sqlparser/expression_rewriting.go [hunk 4]
```diff
@@ -63,31 +79,85 @@ func shouldRewriteDatabaseFunc(in Statement) bool {
...
+//
+//var lastInsertID = funcRewrite{
+//	checkValid: func(f *FuncExpr) error {
+//		if len(f.Exprs) > 0 {
+//			return vterrors.New(vtrpc.Code_UNIMPLEMENTED, "Argument to LAST_INSERT_ID() not supported")
+//		}
+//		return nil
+//	},
+//	bindVarName: LastInsertIDName,
+//}
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/sqlparser/expression_rewriting.go]: oops

**Comment by gitmko0** [comment]
I'm having problem with SQL_CALC_FOUND_ROWS not being Rewrite as well. Is this resolved?
When will it be resolved?

**Pull request merged**